pkg/agent/proxy/instance.go

```diff
@@ -0,0 +1,81 @@
+// Copyright 2020 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxy
+
+import (
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/tools/record"
+
+	"github.com/vmware-tanzu/antrea/pkg/agent/openflow"
+	"github.com/vmware-tanzu/antrea/pkg/agent/proxy/upstream/config"
+	"github.com/vmware-tanzu/antrea/pkg/agent/querier"
+)
+
+const (
+	resyncPeriod  = time.Minute
+	componentName = "antrea-agent-proxy"
+)
+
+type instance struct {
+	serviceConfig   *config.ServiceConfig
+	endpointsConfig *config.EndpointsConfig
+	proxier         *proxier
+
+	once sync.Once
+}
+
+// Instance defines the interface to run the antrea agent proxy.
+type Instance interface {
+	// Run starts the antrea agent proxy, this method should return once it
+	// receives from the stopCh.
+	Run(agentQuerier querier.AgentQuerier, ofClient openflow.Client, stopCh <-chan struct{})
+}
+
+func (i *instance) Run(agentQuerier querier.AgentQuerier, ofClient openflow.Client, stopCh <-chan struct{}) {
+	i.once.Do(func() {
+		i.proxier.agentQuerier = agentQuerier
+		i.proxier.ofClient = ofClient
+
+		go i.serviceConfig.Run(stopCh)
+		go i.endpointsConfig.Run(stopCh)
+
+		i.proxier.stopChan = stopCh
+		i.proxier.SyncLoop()
+	})
+}
+
+func New(hostname string, informerFactory informers.SharedInformerFactory) (Instance, error) {
```

Comment:
It looks like "instance" and "Instance" just creates a proxier and runs it, why is this wrapper needed?

Reply:
I'm used to making an interface for structs that work as a component, because this may be good for test and refactoring.
A second thought is the fields inside the `instance` are actually all internal fields, it is OK to remove the interface here.

If you think this interface is redundant, I'm fine to remove this.

Reply:
I think you can still have an interface for proxier if you want.
The file seems redundant to me.

Reply:
Thanks. Removed the interface.